pkg/helm/controller/reconcile.go

```diff
@@ -141,27 +190,55 @@ func (r HelmOperatorReconciler) Reconcile(request reconcile.Request) (reconcile.
 		previousRelease, updatedRelease, err := manager.UpdateRelease(context.TODO())
 		if err != nil {
 			log.Error(err, "failed to update release")
+			status.SetCondition(types.HelmAppCondition{
+				Type:    types.ConditionReleaseFailed,
+				Status:  types.StatusTrue,
+				Reason:  types.ReasonUpdateError,
+				Message: err.Error(),
+				Release: updatedRelease,
+			})
+			_ = r.updateResourceStatus(o, status)
 			return reconcile.Result{}, err
 		}
+		status.RemoveCondition(types.ConditionReleaseFailed)
+
 		log.Info("Updated release")
 		if log.Enabled() {
 			fmt.Println(diffutil.Diff(previousRelease.GetManifest(), updatedRelease.GetManifest()))
 		}
 		log.V(1).Info("Config values", "values", updatedRelease.GetConfig())
-		status.SetRelease(updatedRelease)
-		status.SetPhase(types.PhaseApplied, types.ReasonApplySuccessful, updatedRelease.GetInfo().GetStatus().GetNotes())
+		status.SetCondition(types.HelmAppCondition{
+			Type:    types.ConditionDeployed,
+			Status:  types.StatusTrue,
+			Reason:  types.ReasonUpdateSuccessful,
+			Message: updatedRelease.GetInfo().GetStatus().GetNotes(),
+			Release: updatedRelease,
+		})
 		err = r.updateResourceStatus(o, status)
 		return reconcile.Result{RequeueAfter: r.ResyncPeriod}, err
 	}
 
 	_, err = manager.ReconcileRelease(context.TODO())
 	if err != nil {
 		log.Error(err, "failed to reconcile release")
+		status.SetCondition(types.HelmAppCondition{
+			Type:    types.ConditionIrreconcilable,
+			Status:  types.StatusTrue,
+			Reason:  types.ReasonReconcileError,
+			Message: err.Error(),
+		})
+		_ = r.updateResourceStatus(o, status)
 		return reconcile.Result{}, err
 	}
+	status.RemoveCondition(types.ConditionIrreconcilable)
 
 	log.Info("Reconciled release")
-	return reconcile.Result{RequeueAfter: r.ResyncPeriod}, nil
+	err = r.updateResourceStatus(o, status)
+	return reconcile.Result{RequeueAfter: r.ResyncPeriod}, err
+}
+
+func (r HelmOperatorReconciler) updateResource(o *unstructured.Unstructured) error {
```

Comment:
this is just my sensibilities, but why do we have this function if it is not abstracting anything but a different function call?

Reply:
Fair question. I was on the fence about this, but I went with it to emphasize the distinction between updating the non-subresources (e.g. metadata, spec, etc.) and the status subresource. So with this we'd have `updateResource()` and `updateResourceStatus()` functions. I wouldn't be opposed to using `r.Client.Update()` directly though.